Source/Core/DolphinQt/Config/Mapping/MappingIndicator.cpp

```diff
@@ -14,174 +14,194 @@
+  const double tv_height = cursor.numeric_settings[cursor.SETTING_HEIGHT]->GetValue();
 
-  // Draw deadzone
-  p.setBrush(Qt::lightGray);
-  p.drawEllipse(centerx - (deadzone / 2), centery - (deadzone / 2), deadzone, deadzone);
+  p.setPen(tv_pen_color);
+  p.setBrush(tv_brush_color);
+  auto gate_polygon = GetPolygonFromRadiusGetter(
+      [&cursor](double ang) { return cursor.GetGateRadiusAtAngle(ang); }, scale);
+  for (auto& pt : gate_polygon)
+  {
+    pt = {pt.x() * tv_width, pt.y() * tv_height + tv_center * center_scale * scale};
+    pt *= tv_scale;
+  }
+  p.drawPolygon(gate_polygon);
 
-  // Draw cursor
-  p.fillRect(curx, cury, 8, 8, Qt::red);
-}
+  // Deadzone.
+  p.setPen(DEADZONE_COLOR);
+  p.setBrush(DEADZONE_BRUSH);
+  p.drawPolygon(GetPolygonFromRadiusGetter(
+      [&cursor](double ang) { return cursor.GetDeadzoneRadiusAtAngle(ang); }, scale));
 
-// Constructs a polygon by querying a radius at varying angles:
-template <typename F>
-QPolygonF GetPolygonFromRadiusGetter(F&& radius_getter, double scale)
-{
-  // A multiple of 8 (octagon) and enough points to be visibly pleasing:
-  constexpr int shape_point_count = 32;
-  QPolygonF shape{shape_point_count};
+  // Input shape.
+  p.setPen(INPUT_SHAPE_PEN);
+  p.setBrush(Qt::NoBrush);
+  p.drawPolygon(GetPolygonFromRadiusGetter(
+      [&cursor](double ang) { return cursor.GetInputRadiusAtAngle(ang); }, scale));
 
-  int p = 0;
-  for (auto& point : shape)
-  {
-    const double angle = MathUtil::TAU * p / shape.size();
-    const double radius = radius_getter(angle) * scale;
+  // Raw stick position.
+  p.setPen(Qt::NoPen);
+  p.setBrush(RAW_INPUT_COLOR);
+  p.drawEllipse(QPointF{raw_coord.x, raw_coord.y} * scale, INPUT_DOT_RADIUS, INPUT_DOT_RADIUS);
 
-    point = {std::cos(angle) * radius, std::sin(angle) * radius};
-    ++p;
+  // Adjusted cursor position if not hidden:
+  if (adj_coord.x < 10000)
+  {
+    p.setPen(Qt::NoPen);
+    p.setBrush(ADJ_INPUT_COLOR);
+    const QPointF pt(adj_coord.x / 2.0, (adj_coord.y - tv_center) / 2.0 + tv_center * center_scale);
+    p.drawEllipse(pt * scale * tv_scale, INPUT_DOT_RADIUS, INPUT_DOT_RADIUS);
   }
-
-  return shape;
 }
 
-void MappingIndicator::DrawStick()
+void MappingIndicator::DrawReshapableInput(ControllerEmu::ReshapableInput& stick)
 {
-  // Make the c-stick yellow:
+  // Some hacks for pretty colors:
   const bool is_c_stick = m_group->name == "C-Stick";
-  const QColor gate_brush_color = is_c_stick ? Qt::yellow : Qt::lightGray;
-  const QColor gate_pen_color = gate_brush_color.darker(125);
+  const bool is_tilt = m_group->name == "Tilt";
+  const bool is_cursor = m_group->name == "IR";
+
+  QColor gate_brush_color = Qt::lightGray;
 
-  auto& stick = *static_cast<ControllerEmu::AnalogStick*>(m_group);
+  if (is_c_stick)
+    gate_brush_color = Qt::yellow;
+  else if (is_tilt)
+    gate_brush_color = 0xa2d9ce;
+  else if (is_cursor)
+    gate_brush_color = 0xaed6f1;
```

Comment:
Any reason you don't put those up there with the other `const` colors (and other stuff) at the top of the file?

Reply:
I could. That else-if-ladder is pretty hacky there already. I figured I'd keep all the color hackery in one spot. (Hacky UI code is allowed. :P)

Reply:
I moved them to the top now with the others. :)